File: engine.py
"""
Python port of static/engine.js — must produce bit-identical results so the
server can replay a recorded game and compute the authoritative score.

The 32-bit integer arithmetic mirrors JavaScript's |0, >>>, ^, | and
Math.imul semantics exactly (verified by a cross-check against the JS engine).
"""

COLS, ROWS = 10, 20

SHAPES = {
    "I": [[0, 0, 0, 0], [1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0]],
    "O": [[2, 2], [2, 2]],
    "T": [[0, 3, 0], [3, 3, 3], [0, 0, 0]],
    "S": [[0, 4, 4], [4, 4, 0], [0, 0, 0]],
    "Z": [[5, 5, 0], [0, 5, 5], [0, 0, 0]],
    "J": [[6, 0, 0], [6, 6, 6], [0, 0, 0]],
    "L": [[0, 0, 7], [7, 7, 7], [0, 0, 0]],
}
TYPES = ["I", "O", "T", "S", "Z", "J", "L"]
LINE_SCORES = [0, 40, 100, 300, 1200]
KICKS = [0, -1, 1, -2, 2]
ACTIONS = ("L", "R", "ROT", "SOFT", "HARD")

_MASK = 0xFFFFFFFF


def _u32(x):
    return x & _MASK


def _i32(x):
    x &= _MASK
    return x - 0x100000000 if x & 0x80000000 else x


def _imul(a, b):
    """JS Math.imul: signed 32-bit multiply, low 32 bits."""
    return _i32((_i32(a) * _i32(b)) & _MASK)


def _ushr(x, n):
    """JS >>> : unsigned right shift on a 32-bit value."""
    return _u32(x) >> n


def _xor(x, y):
    return _i32(_i32(x) ^ _i32(y))


def _or(x, y):
    return _i32(_i32(x) | _i32(y))


def gravity_ticks(level):
    return max(2, 32 - (level - 1) * 3)


class _Rng:
    """mulberry32 returning unsigned 32-bit integers."""

    def __init__(self, seed):
        self.a = _i32(seed)

    def next(self):
        a = _i32(self.a)
        a = _i32(a + 0x6D2B79F5)
        self.a = a
        t = _imul(_xor(a, _ushr(a, 15)), _or(1, a))
        t = _u32(_xor(_i32(t + _imul(_xor(t, _ushr(t, 7)), _or(61, t))), t))
        return _u32(_xor(t, _ushr(t, 14)))


def _shape_matrix(t):
    return [row[:] for row in SHAPES[t]]


def _rotate_cw(m):
    n = len(m)
    out = [[0] * n for _ in range(n)]
    for y in range(n):
        for x in range(n):
            out[x][n - 1 - y] = m[y][x]
    return out


class Engine:
    def __init__(self, seed):
        self.seed = _u32(seed)
        self.rng = _Rng(self.seed)
        self.bag = []
        self.grid = [[0] * COLS for _ in range(ROWS)]
        self.score = 0
        self.lines = 0
        self.level = 1
        self.gravity_counter = 0
        self.game_over = False
        self.current = None
        self.next_type = None
        self._started = False

    def _draw_type(self):
        if not self.bag:
            self.bag = TYPES[:]
            for i in range(len(self.bag) - 1, 0, -1):
                j = self.rng.next() % (i + 1)
                self.bag[i], self.bag[j] = self.bag[j], self.bag[i]
        return self.bag.pop()

    def start(self):
        if self._started:
            return
        self._started = True
        self.next_type = self._draw_type()
        self._spawn()

    def _make_piece(self, t):
        matrix = _shape_matrix(t)
        return {"type": t, "matrix": matrix, "x": (COLS - len(matrix[0])) // 2, "y": 0}

    def _spawn(self):
        self.current = self._make_piece(self.next_type)
        self.next_type = self._draw_type()
        if self._collides(self.current):
            self.game_over = True

    def _collides(self, p):
        m = p["matrix"]
        for y in range(len(m)):
            for x in range(len(m[y])):
                if not m[y][x]:
                    continue
                bx, by = p["x"] + x, p["y"] + y
                if bx < 0 or bx >= COLS or by >= ROWS:
                    return True
                if by >= 0 and self.grid[by][bx]:
                    return True
        return False

    def _merge(self, p):
        m = p["matrix"]
        for y in range(len(m)):
            for x in range(len(m[y])):
                if m[y][x] and p["y"] + y >= 0:
                    self.grid[p["y"] + y][p["x"] + x] = m[y][x]

    def _clear_lines(self):
        cleared = 0
        y = ROWS - 1
        while y >= 0:
            if all(self.grid[y][x] for x in range(COLS)):
                del self.grid[y]
                self.grid.insert(0, [0] * COLS)
                cleared += 1
                y += 1
            y -= 1
        if cleared:
            self.score += LINE_SCORES[cleared] * self.level
            self.lines += cleared
            self.level = self.lines // 10 + 1

    def _lock(self):
        self._merge(self.current)
        self._clear_lines()
        self.gravity_counter = 0
        self._spawn()

    def apply_action(self, action):
        if self.game_over or self.current is None:
            return
        p = self.current
        if action == "L":
            p["x"] -= 1
            if self._collides(p):
                p["x"] += 1
        elif action == "R":
            p["x"] += 1
            if self._collides(p):
                p["x"] -= 1
        elif action == "ROT":
            prev_matrix, px = p["matrix"], p["x"]
            p["matrix"] = _rotate_cw(p["matrix"])
            ok = False
            for k in KICKS:
                p["x"] = px + k
                if not self._collides(p):
                    ok = True
                    break
            if not ok:
                p["matrix"], p["x"] = prev_matrix, px
        elif action == "SOFT":
            p["y"] += 1
            if self._collides(p):
                p["y"] -= 1
                self._lock()
            else:
                self.score += 1
                self.gravity_counter = 0
        elif action == "HARD":
            dist = 0
            while not self._collides(p):
                p["y"] += 1
                dist += 1
            p["y"] -= 1
            dist -= 1
            if dist > 0:
                self.score += dist * 2
            self._lock()

    def gravity_step(self):
        if self.game_over or self.current is None:
            return
        self.gravity_counter += 1
        if self.gravity_counter >= gravity_ticks(self.level):
            self.gravity_counter = 0
            p = self.current
            p["y"] += 1
            if self._collides(p):
                p["y"] -= 1
                self._lock()


def simulate(seed, inputs, max_ticks=5_000_000):
    """Replay a recorded game. inputs: iterable of {"tick": int, "action": str}."""
    e = Engine(seed)
    e.start()
    if e.game_over:
        return {"score": e.score, "lines": e.lines, "level": e.level,
                "gameOver": True, "ticks": 0}

    by_tick = {}
    last_tick = 0
    for it in inputs:
        try:
            t = int(it["tick"])
            a = it["action"]
        except (KeyError, TypeError, ValueError):
            continue
        if a not in ACTIONS or t < 0:
            continue
        if t > last_tick:
            last_tick = t
        by_tick.setdefault(t, []).append(a)

    end_tick = last_tick + 100000
    tick = 0
    while not e.game_over and tick <= end_tick and tick <= max_ticks:
        for a in by_tick.get(tick, ()):
            e.apply_action(a)
            if e.game_over:
                break
        if e.game_over:
            break
        e.gravity_step()
        tick += 1

    return {"score": e.score, "lines": e.lines, "level": e.level,
            "gameOver": e.game_over, "ticks": tick}

File: static/engine.js
/*
 * Deterministic Tetris engine — no DOM, no timers.
 * Given the same (seed, inputs) it always produces the same score, so the
 * server can replay a game and compute the score itself.
 *
 * Works both in the browser (as global `TetrisEngine`) and under Node
 * (`require`) so it can be cross-checked against the Python port.
 */
(function (root, factory) {
  const api = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  else root.TetrisEngine = api;
})(typeof globalThis !== "undefined" ? globalThis : this, function () {
  "use strict";

  const COLS = 10, ROWS = 20;

  const SHAPES = {
    I: [[0, 0, 0, 0], [1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0]],
    O: [[2, 2], [2, 2]],
    T: [[0, 3, 0], [3, 3, 3], [0, 0, 0]],
    S: [[0, 4, 4], [4, 4, 0], [0, 0, 0]],
    Z: [[5, 5, 0], [0, 5, 5], [0, 0, 0]],
    J: [[6, 0, 0], [6, 6, 6], [0, 0, 0]],
    L: [[0, 0, 7], [7, 7, 7], [0, 0, 0]],
  };
  // Fixed order — the bag shuffle depends on it, so it must match the port.
  const TYPES = ["I", "O", "T", "S", "Z", "J", "L"];
  const LINE_SCORES = [0, 40, 100, 300, 1200];
  const KICKS = [0, -1, 1, -2, 2];
  const ACTIONS = ["L", "R", "ROT", "SOFT", "HARD"];

  function gravityTicks(level) {
    return Math.max(2, 32 - (level - 1) * 3);
  }

  // mulberry32, returning an unsigned 32-bit integer (no float division).
  function makeRng(seed) {
    let a = seed >>> 0;
    return function () {
      a |= 0;
      a = (a + 0x6d2b79f5) | 0;
      let t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = ((t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t) >>> 0;
      return (t ^ (t >>> 14)) >>> 0;
    };
  }

  const shapeMatrix = (type) => SHAPES[type].map((row) => row.slice());

  function rotateCW(m) {
    const n = m.length;
    const out = [];
    for (let i = 0; i < n; i++) out.push(new Array(n).fill(0));
    for (let y = 0; y < n; y++) for (let x = 0; x < n; x++) out[x][n - 1 - y] = m[y][x];
    return out;
  }

  class Engine {
    constructor(seed) {
      this.seed = seed >>> 0;
      this.rng = makeRng(this.seed);
      this.bag = [];
      this.grid = [];
      for (let y = 0; y < ROWS; y++) this.grid.push(new Array(COLS).fill(0));
      this.score = 0;
      this.lines = 0;
      this.level = 1;
      this.gravityCounter = 0;
      this.gameOver = false;
      this.current = null;
      this.nextType = null;
      this._started = false;
    }

    _drawType() {
      if (this.bag.length === 0) {
        this.bag = TYPES.slice();
        for (let i = this.bag.length - 1; i > 0; i--) {
          const j = this.rng() % (i + 1);
          const tmp = this.bag[i];
          this.bag[i] = this.bag[j];
          this.bag[j] = tmp;
        }
      }
      return this.bag.pop();
    }

    start() {
      if (this._started) return;
      this._started = true;
      this.nextType = this._drawType();
      this._spawn();
    }

    _makePiece(type) {
      const matrix = shapeMatrix(type);
      return { type, matrix, x: Math.floor((COLS - matrix[0].length) / 2), y: 0 };
    }

    _spawn() {
      this.current = this._makePiece(this.nextType);
      this.nextType = this._drawType();
      if (this._collides(this.current)) this.gameOver = true;
    }

    _collides(p) {
      const m = p.matrix;
      for (let y = 0; y < m.length; y++) {
        for (let x = 0; x < m[y].length; x++) {
          if (!m[y][x]) continue;
          const bx = p.x + x, by = p.y + y;
          if (bx < 0 || bx >= COLS || by >= ROWS) return true;
          if (by >= 0 && this.grid[by][bx]) return true;
        }
      }
      return false;
    }

    _merge(p) {
      for (let y = 0; y < p.matrix.length; y++) {
        for (let x = 0; x < p.matrix[y].length; x++) {
          if (p.matrix[y][x] && p.y + y >= 0) this.grid[p.y + y][p.x + x] = p.matrix[y][x];
        }
      }
    }

    _clearLines() {
      let cleared = 0;
      for (let y = ROWS - 1; y >= 0; y--) {
        let full = true;
        for (let x = 0; x < COLS; x++) if (!this.grid[y][x]) { full = false; break; }
        if (full) {
          this.grid.splice(y, 1);
          this.grid.unshift(new Array(COLS).fill(0));
          cleared++;
          y++;
        }
      }
      if (cleared) {
        this.score += LINE_SCORES[cleared] * this.level;
        this.lines += cleared;
        this.level = Math.floor(this.lines / 10) + 1;
      }
    }

    _lock() {
      this._merge(this.current);
      this._clearLines();
      this.gravityCounter = 0;
      this._spawn();
    }

    // Highest y the current piece can drop to (for ghost rendering).
    ghostY() {
      if (!this.current) return 0;
      const p = { matrix: this.current.matrix, x: this.current.x, y: this.current.y };
      while (!this._collides({ matrix: p.matrix, x: p.x, y: p.y + 1 })) p.y++;
      return p.y;
    }

    applyAction(action) {
      if (this.gameOver || !this.current) return;
      const p = this.current;
      if (action === "L") {
        p.x--; if (this._collides(p)) p.x++;
      } else if (action === "R") {
        p.x++; if (this._collides(p)) p.x--;
      } else if (action === "ROT") {
        const prevMatrix = p.matrix, px = p.x;
        p.matrix = rotateCW(p.matrix);
        let ok = false;
        for (let i = 0; i < KICKS.length; i++) {
          p.x = px + KICKS[i];
          if (!this._collides(p)) { ok = true; break; }
        }
        if (!ok) { p.matrix = prevMatrix; p.x = px; }
      } else if (action === "SOFT") {
        p.y++;
        if (this._collides(p)) { p.y--; this._lock(); }
        else { this.score += 1; this.gravityCounter = 0; }
      } else if (action === "HARD") {
        let dist = 0;
        while (!this._collides(p)) { p.y++; dist++; }
        p.y--; dist--;
        if (dist > 0) this.score += dist * 2;
        this._lock();
      }
    }

    gravityStep() {
      if (this.gameOver || !this.current) return;
      this.gravityCounter++;
      if (this.gravityCounter >= gravityTicks(this.level)) {
        this.gravityCounter = 0;
        const p = this.current;
        p.y++;
        if (this._collides(p)) { p.y--; this._lock(); }
      }
    }
  }

  // Replay a recorded game. inputs: [{ tick, action }, ...]
  function simulate(seed, inputs, maxTicks) {
    maxTicks = maxTicks || 5000000;
    const e = new Engine(seed);
    e.start();
    if (e.gameOver) {
      return { score: e.score, lines: e.lines, level: e.level, gameOver: true, ticks: 0 };
    }

    const byTick = new Map();
    let lastTick = 0;
    for (const it of inputs) {
      const t = it.tick | 0;
      const a = it.action;
      if (ACTIONS.indexOf(a) === -1) continue; // ignore unknown actions
      if (t > lastTick) lastTick = t;
      if (!byTick.has(t)) byTick.set(t, []);
      byTick.get(t).push(a);
    }

    const endTick = lastTick + 100000; // let gravity finish after the last input
    let tick = 0;
    while (!e.gameOver && tick <= endTick && tick <= maxTicks) {
      const acts = byTick.get(tick);
      if (acts) {
        for (const a of acts) { e.applyAction(a); if (e.gameOver) break; }
      }
      if (e.gameOver) break;
      e.gravityStep();
      tick++;
    }
    return { score: e.score, lines: e.lines, level: e.level, gameOver: e.gameOver, ticks: tick };
  }

  return { Engine, simulate, gravityTicks, makeRng, COLS, ROWS, TYPES, SHAPES, ACTIONS };
});
